src/lib/yacc/grammar.rs

```diff
@@ -465,20 +466,36 @@ impl<'a> SentenceGenerator<'a> {
         where F: Fn(TIdx) -> u64
     {
         let nt_min_costs = nonterm_min_costs(grm, &term_cost);
+        let nt_max_costs = nonterm_max_costs(grm, &term_cost);
         let mut term_costs = Vec::with_capacity(grm.terms_len());
         for i in 0..grm.terms_len() {
             term_costs.push(term_cost(TIdx::from(i)));
         }
-        SentenceGenerator{grm, term_costs, nonterm_min_costs: nt_min_costs}
+        SentenceGenerator{grm,
+                          term_costs,
+                          nonterm_min_costs: nt_min_costs,
+                          nonterm_max_costs: nt_max_costs}
     }
 
-    /// What is the cost of a minimal sentence for the non-terminal `nonterm_idx`. Note that,
+    /// What is the cost of a minimal sentence for the non-terminal `nonterm_idx`? Note that,
     /// unlike `min_sentence`, this function does not actually *build* a sentence and it is thus
     /// much faster.
     pub fn min_sentence_cost(&self, nonterm_idx:NTIdx) -> u64 {
         self.nonterm_min_costs[usize::from(nonterm_idx)]
     }
 
+    /// What is the cost of a maximal sentence for the non-terminal `nonterm_idx`? Non-terminals
+    /// which can generate sentences of unbounded length return None; non-terminals which can only
+    /// generate maximal strings of a finite length return a `Some(u64)`.
+    pub fn max_sentence_cost(&self, nonterm_idx:NTIdx) -> Option<u64> {
+        let v = self.nonterm_max_costs[usize::from(nonterm_idx)];
+        if v == u64::max_value() {
```

Comment:
I guess we don't want to (or can't) mix `None` with `u64` in `costs` in `nonterm_max_costs`? Otherwise we could get rid of `max_value` in both functions.

Reply:
We could, but then we double storage costs to 128 bits :/